### PR TITLE
Implementation of `Dataframe.merge/join` by shuffle

### LINKED ISSUE #571: Support join operator for DataFrame
**Is your feature request related to a problem? Please describe.**
Implement join operator for joining columns with other DataFrame either on index or on a key column. This is an operator rely on shuffle and need to note performance issues.

### PR DESCRIPTION
**Pull request by sighingnow**
## What do these changes do?

A proof-of-concept implementation of `DataFrame.merge/join` (in pandas the `join` is implemented using `merge`).

The `merge` is implemented by shuffle on the keys that to be used to merge.

## Related issue number

Resolves #571

### CODE REVIEW

**Comment by qinxuye** [comment]
Exciting work! Just so quick beyond my expectation.

BTW, the expression and execution will no longer be separated described in #453, thus this PR may need to be refactored soon to follow the new architecture.

**Comment by qinxuye** [comment]
And just to confirm that we will finally reach a solid implementation for this PR, right? I hope current title just illustrates the WIP status.

**Comment by sighingnow** [comment]
> And just to confirm that we will finally reach a solid implementation for this PR, right?

Yes. I will revise the patch after the expression/execution refactor done, and I will continue to add more detailed optimization (and may take ideas from mask/modin/pandas) util the implementation reach a solid state.

**Comment by qinxuye** [comment]
Are you interested in the GPU part of the merge/join for this PR as well? I think we could provide the ability based on `cudf` to reach the goal. If not, we can focus on CPU only for this PR.

**Comment by sighingnow** [comment]
> I think we could provide the ability based on cudf to reach the goal. 

I have no experience with the `cudf` package but I think it can just be used as `pandas`, (such as `numpy` and `cupy`), though maybe special optimization is needed for `cupy`.

I'm interested in this task and could place it on my TODO after we finish ops proposed in #495.

**Comment by sighingnow** [comment]
Update for this PR (just FYI):

+ Implement a basic `set_index` on dataframe, currently only support set an existing column as `index` and don't verify the integrity. (the default value of `verify_integrity` is `False` in pandas).

**Comment by qinxuye** [comment]
> > I think we could provide the ability based on cudf to reach the goal.
> 
> I have no experience with the `cudf` package but I think it can just be used as `pandas`, (such as `numpy` and `cupy`), though maybe special optimization is needed for `cupy`.
> 
> I'm interested in this task and could place it on my TODO after we finish ops proposed in #495.

Great.

**Comment by qinxuye** [comment]
We are considering to use Azure pipelines for ci. For now, in order to address the pyarrow issue for appveyor, you can refer to https://github.com/mars-project/mars/pull/573/commits/817d27c53271710448806269d444a1dc619e9abc to limit the pyarrow version as a workaround.

**Comment by qinxuye** [comment]
Huge refactor has been merged, thus this PR could be rebased.

**Comment by sighingnow** [comment]
Update:

Now this pr focus on shuffle join, any other unrelated snippets have been removed. Exhaustive test for the correctness with different combination of parameters has been added as well. I think now this pr is ready for review.

For reviewers, you may need to read through those three notes in `test_merge_execution.py`.

File: mars/dataframe/merge/tests/test_merge_execution.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        mdf2 = from_pandas(df2, chunk_size=2)
+
+        # Note [Index of Merge]
+        #
+        # When `left_index` and `right_index` of `merge` is both false, pandas will generate an RangeIndex to
+        # the final result dataframe.
+        #
+        # We chunked the `left` and `right` dataframe, thus every result chunk will have its own RangeIndex.
+        # When they are contenated we don't generate a new RangeIndex for the result, thus we cannot obtain the
+        # same index value with pandas. But we guarantee that the content of dataframe is correct.
```
Comment (qinxuye) [inline mars/dataframe/merge/tests/test_merge_execution.py]: Yeah, this is a problem, we can create a new op to do this, that is, for a chunk, calculating the max index, and recalculate the index according to all previous ones.

For some situation, this op could be skippable, for instance, `to_tensor`.

I think this op could be added after this PR gotta merged.

File: mars/dataframe/utils.py [hunk 2]
```diff
@@ -12,33 +12,48 @@
...
-import pandas as pd
-from mars.lib.mmh3 import hash as mmh_hash
-
 import itertools
 import operator
 import functools
 
+import numpy as np
+import pandas as pd
+from mars.lib.mmh3 import hash as mmh_hash
```
Comment (qinxuye) [inline mars/dataframe/utils.py]: Could be from `..lib.mmh3 import hash as mmh_hash` using a relative import, the old one must be wrong and not been discovered before.

File: mars/opcodes.py [hunk 3]
```diff
@@ -288,5 +288,8 @@
 RESHAPE_REDUCE = 2003
 DATAFRAME_INDEX_ALIGN_MAP = 2004
 DATAFRAME_INDEX_ALIGN_REDUCE = 2005
+SHUFFLE_MERGE = 2010
```
Comment (qinxuye) [inline mars/opcodes.py]: Maybe call it with a prefix `DATAFRAME_` cuz it's used only for dataframe?

File: mars/dataframe/utils.py [hunk 4]
```diff
@@ -12,33 +12,48 @@
...
 
     f = functools.partial(func, size=size)
-    idx_to_grouped = dict(index.groupby(index.map(f)).items())
+    idx_to_grouped = index.groupby(index.map(f))
     return [idx_to_grouped.get(i, list()) for i in range(size)]
 
 
+def hash_dataframe_on(df, on, size):
+    if on is None:
+        hashed_label = pd.util.hash_pandas_object(df.index, categorize=False)
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply mars/dataframe/utils.py]: Here I use `pd.util.hash_pandas_object` just for convenience. For `mmh3 ` when `merge` on only one column, we need to hash on `pd.Index` and when `merge` on multiple columns we need to concatenate multiple values to a `bytes` and then do `hash`.

I haven't do any quantitive analysis about the performance and the hash diversity.

File: mars/dataframe/merge/merge.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+                                     nsplits=tuple(tuple(ns) for ns in nsplits),
+                                     chunks=out_chunks, dtypes=df.dtypes,
+                                     index_value=df.index_value, columns_value=df.columns)
+
+    @classmethod
+    def execute(cls, ctx, op):
+        chunk = op.outputs[0]
+        left, right = ctx[op.inputs[0].key], ctx[op.inputs[1].key]
+        r = left.merge(right, how=op.how, on=op.on,
+                                    left_on=op.left_on, right_on=op.right_on,
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply mars/dataframe/merge/merge.py]: Thanks! Fixed now.

File: mars/opcodes.py (hunk 3 above)
Comment (sighingnow) [reply mars/opcodes.py]: Fixed. `DATAFRAME_` prefix has been added.

File: mars/dataframe/merge/tests/test_merge_execution.py (hunk 1 above)
Comment (sighingnow) [reply mars/dataframe/merge/tests/test_merge_execution.py]: Maybe we could use some kind of `Index` called `UnknownIndex` and when build concat dataframe we generate a new `RangeIndex` for the result.

File: mars/dataframe/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (qinxuye) [reply mars/dataframe/utils.py]: Yeah, that makes sense, we can do some experiments later, current implementation looks fine.

**Comment by qinxuye** [comment]
Shall we add a simple test in test_cluster, as I known, the shuffle may be different in the distributed setting.

**Comment by sighingnow** [comment]
> Shall we add a simple test in test_cluster, as I known, the shuffle may be different in the distributed setting.

Done.

**Comment by qinxuye** [comment]
`DataFrame.sum` gotta merged... Conflict could be resolved.

File: mars/dataframe/merge/merge.py [hunk 6]
```diff
@@ -246,6 +246,9 @@ def tile(cls, op):
     def execute(cls, ctx, op):
         chunk = op.outputs[0]
         left, right = ctx[op.inputs[0].key], ctx[op.inputs[1].key]
+        # workaround for: https://github.com/pandas-dev/pandas/issues/27943
+        left = left.copy(deep=True)
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py]: How about copy when error happens?

Comment (sighingnow) [reply mars/dataframe/merge/merge.py]: Agree.

**Pull request merged**